Feature/dictionary

varvarkina opened the pull request:
Разработала консольное приложение Dictionary на языке C#, выполняющее роль словаря.

varvarkina committed:
Initialize repository and add gitignore

varvarkina committed:
Complete dictionary homework

VladDotMe commented on Dictionary/Dictionary/Program.cs:
```diff
@@ -0,0 +1,181 @@
+
+static void PrintWelcomeMessage()
+{
+    Console.Write( @"Чтобы воспользоваться словарем, нажмите ""Enter""" );
```
Есть какая-то причина почему используешь Console.Write вместо WriteLine()?

varvarkina replied:
Чтобы из-за переноса на новую строку и нажатия Enter не создавалась лишняя пустая строка

VladDotMe commented on Dictionary/Dictionary/Program.cs:
```diff
@@ -0,0 +1,181 @@
+            AddWords();
+            break;
+        default:
+            Console.WriteLine( "Unknown command" );
```
Хорошее правило весь ввод/вывод делать на одном языке, поскольк у тебя все на русском то и это сообщение должно быть на русском.

VladDotMe commented on Dictionary/Dictionary/Program.cs:
```diff
@@ -0,0 +1,181 @@
+
+static void RusToEng()
+{
+    Dictionary<string, string> dictionary = BuildDictionary();
```
Ты сейчас получается читаешь файл (собираешь заново словарь в памяти) каждый раз при выполнении любого действия, мне кажется это не очень хорошо, представь что у тебя там будет миллион записей и если ты будешь каждый раз считывать файл - будут проблемы) 

Нужно придумать другое решение :)

VladDotMe commented on Dictionary/Dictionary/Program.cs:
```diff
@@ -0,0 +1,181 @@
+    Console.WriteLine( "[3] Добавить новое слово в словарь" );
+}
+
+static Dictionary<string, string> BuildDictionary()
```
У меня падает ошибка т.к. нет файла Dictionary.txt, надо тоже этот кейс обработать

varvarkina committed:
Fix bug, improve the performance, change the style, add new feature

Fix: create the "Dictionary.txt" file if it doesn't exist
Perf: use BuildDictionary()  twice: at the top and in AddWords() after changing dictionary
Style: Make the output message in English, add constant
Feat: add .ToLower() method

VladDotMe commented on Dictionary/Dictionary/Program.cs:
```diff
@@ -0,0 +1,186 @@
+static Dictionary<string, string> BuildDictionary()
+{
+    var dictionary = new Dictionary<string, string>();
+    string path = "C:/study/tl_practice_2024/Dictionary/Dictionary/bin/Debug/net8.0/Dictionary.txt";
```
абсолютный путь нельзя использовать. У меня, например, такого пути нет и не будет работать. Нужно использовать относительный.

VladDotMe commented on Dictionary/Dictionary/Program.cs:
```diff
@@ -0,0 +1,186 @@
+{
+    var dictionary = new Dictionary<string, string>();
+    string path = "C:/study/tl_practice_2024/Dictionary/Dictionary/bin/Debug/net8.0/Dictionary.txt";
+    if ( !File.Exists( path ) )
```
здесь нет необходимости создавать файл при его отсутствии т.к. мы хотим из него считывать, лучше просто return dictionary

VladDotMe commented on Dictionary/Dictionary/Program.cs:
```diff
@@ -0,0 +1,186 @@
+                "Введите другой перевод этого слова на английский: " );
+            engWord = EngProcess();
+        }
+        using ( StreamWriter writer = new StreamWriter( path, true ) )
```
А вот здесь можно переделать чтобы при добавлении нового текста при необходимости (если файл отсутствует) - он создавался.  Подсказка: можно заиспользовать File.AppendText( path)

VladDotMe commented:
А еще не хватает функции выхода из программы)

VladDotMe commented on Dictionary/Dictionary/Program.cs:
```diff
@@ -0,0 +1,186 @@
+
+static string RusProcess()
+{
+    string word = Console.ReadLine().ToLower();
```
что будет если Console.ReadLine() вернет ошибку? (например если ввести Ctrl+Z, Enter)

VladDotMe commented on Dictionary/Dictionary/Program.cs:
```diff
@@ -0,0 +1,186 @@
+        }
+        Console.Write( "Ошибка! Введите слово, состоящее только из букв " +
+            "русского алфавита, тире и апострофа.\nРусский: " );
+        word = Console.ReadLine();
```
а здесь не приводишь word к нижнему регистру

VladDotMe commented on Dictionary/Dictionary/Program.cs:
```diff
@@ -0,0 +1,186 @@
+
+static string EngProcess()
+{
+    string word = Console.ReadLine().ToLower();
```
здесь и ниже аналогично как в RusProcess

varvarkina committed:
Fix bugs, improve the performance, add new features

Fix: - fix Ctrl+Z input bug
      - use relative path
Perf: create file "Dictionary.txt" if necessary with File.AppendText( path )
Feat: - add exit function
        - add .ToLower() method

VladDotMe commented on Dictionary/Dictionary/Program.cs:
```diff
@@ -0,0 +1,190 @@
+        count++;
+        if ( count == tryLimit )
+        {
+            throw new Exception();
```
Выбрасывать Exception и потом отлавливать и ничего не делать - не очень подход. 
Лучше вернуть null или bool по примеру C# дефолтных методов типа TryParse
Как пример:
https://stackoverflow.com/questions/67468421/best-practices-for-when-to-use-the-try-pattern-in-c-sharp

Но можешь непариться, просто верунть null и потом проверить что строка не null

VladDotMe commented on Dictionary/Dictionary/Program.cs:
```diff
@@ -0,0 +1,190 @@
+        Console.WriteLine( "Слово успешно добавлено в словарь!" );
+        dictionary = BuildDictionary();
+    }
+    catch ( Exception )
```
Здесь аналогично, лучше не генерить пустые Exception а просто проверять на null.

VladDotMe commented on Dictionary/Dictionary/Program.cs:
```diff
@@ -0,0 +1,190 @@
+    try
+    {
+        string rusWord = RusProcess();
+        while ( dictionary.ContainsKey( rusWord ) )
```
Стоит делить сложный код новыми строками на логические блоки, т.е.:
        string path = "Dictionary.txt";

        Console.Write( "Введите слово на русском языке: " );
        string rusWord = RusProcess();
        while ( dictionary.ContainsKey( rusWord ) )
        {
            Console.Write( "Это слово уже есть в словаре!\n" +
                "Введите другое слово на русском языке: " );
            rusWord = RusProcess();
        }

        Console.Write( "Введите перевод этого слова на английский: " );
        string engWord = EngProcess();
        while ( dictionary.ContainsValue( engWord ) )
        {
            Console.Write( "Это слово уже есть в словаре!\n" +
                "Введите другой перевод этого слова на английский: " );
            engWord = EngProcess();
        }

        using ( StreamWriter writer = File.AppendText( path ) )
        {
            writer.WriteLine( $"{rusWord} {engWord}" );
        }
        Console.WriteLine( "Слово успешно добавлено в словарь!" );
        
        dictionary = BuildDictionary();

VladDotMe commented on Dictionary/Dictionary/Program.cs:
```diff
@@ -0,0 +1,190 @@
+            "латинского алфавита, тире и апострофа.\nАнглийский: " );
+        word = Console.ReadLine();
+    }
+    return word.ToLower();
```
и перед return обычно всегда новая строка

varvarkina committed:
Return null instead of throwing exception, divide into logical blocks with bold lines
